d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# CSE Coding Assignment
## Instructions

- Please answer all questions
- You can use any language you wish (e.g. Python, Scala, SQL...)
- Several Markdown cells require completion. Please edit the Markdown cells to include your answer.
- Your final notebook should compile without errors when you click "Run All"

**Please do not publish questions. This is a confidential assignment.**

### Creating a Cluster

You will need to create a Databricks Cluster. More information on this process is available here: https://docs.databricks.com/user-guide/clusters/create.html

## Getting Started

**REQUIRED:** Run the following cells exactly as written to retrieve the necessary Coding Assignment Data Sets from Amazon S3.

In [0]:
%sh curl --remote-name-all 'https://files.training.databricks.com/assessments/cse-take-home/{covertype,kafka,treecover,u.data,u.item}.csv'

[1/5]: https://files.training.databricks.com/assessments/cse-take-home/covertype.csv --> covertype.csv
 % Total % Received % Xferd Average Speed Time Time Time Current
 Dload Upload Total Spent Left Speed

 0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0
100 133 100 133 0 0 293 0 --:--:-- --:--:-- --:--:-- 293

[2/5]: https://files.training.databricks.com/assessments/cse-take-home/kafka.csv --> kafka.csv

 46 135k 46 65008 0 0 910k 0 --:--:-- --:--:-- --:--:-- 910k
100 135k 100 135k 0 0 1764k 0 --:--:-- --:--:-- --:--:-- 10.0M

[3/5]: https://files.training.databricks.com/assessments/cse-take-home/treecover.csv --> treecover.csv

100 642k 100 642k 0 0 8674k 0 --:--:-- --:--:-- --:--:-- 8674k

[4/5]: https://files.training.databricks.com/assessments/cse-take-home/u.data.csv --> u.data.csv

100 1932k 100 1932k 0 0 26.2M 0 --:--:-- --:--:-- --:--:-- 26.2M

[5/5]: https://files.training.databricks.com/assessments/cse-take-home/u.item.csv --> u.item.csv

100 230k 100 230k 0 0 4152k 0 --:--:-- --:--:-- --:--:-- 4152k
--_curl_--https://files.training.databricks.com/assessments/cse-take-home/covertype.csv
--_curl_--https://files.training.databricks.com/assessments/cse-take-home/kafka.csv
--_curl_--https://files.training.databricks.com/assessments/cse-take-home/treecover.csv
--_curl_--https://files.training.databricks.com/assessments/cse-take-home/u.data.csv
--_curl_--https://files.training.databricks.com/assessments/cse-take-home/u.item.csv

In [0]:
%sh cat u.data.csv

196	242	3	881250949
186	302	3	891717742
22	377	1	878887116
244	51	2	880606923
166	346	1	886397596
298	474	4	884182806
115	265	2	881171488
253	465	5	891628467
305	451	3	886324817
6	86	3	883603013
62	257	2	879372434
286	1014	5	879781125
200	222	5	876042340
210	40	3	891035994
224	29	3	888104457
303	785	3	879485318
122	387	5	879270459
194	274	2	879539794
291	1042	4	874834944
234	1184	2	892079237
119	392	4	886176814
167	486	4	892738452
299	144	4	877881320
291	118	2	874833878
308	1	4	887736532
95	546	2	879196566
38	95	5	892430094
102	768	2	883748450
63	277	4	875747401
160	234	5	876861185
50	246	3	877052329
301	98	4	882075827
225	193	4	879539727
290	88	4	880731963
97	194	3	884238860
157	274	4	886890835
181	1081	1	878962623
278	603	5	891295330
276	796	1	874791932
7	32	4	891350932
10	16	4	877888877
284	304	4	885329322
201	979	2	884114233
276	564	3	874791805
287	327	5	875333916
246	201	5	884921594
242	1137	5	879741196
249	241	5	879641194
99	4	5	886519097
178	332	3	882823437
251	100	4	886271884
81	432	2	876535131
260	322	4	890618898
25	181	5	885853415
59	196	5	888205088
72	679	2	880037164
87	384	4	879877127
290	143	5	880474293
42	423	5	881107687
292	515	4	881103977
115	20	3	881171009
20	288	1	879667584
201	219	4	884112673
13	526	3	882141053
246	919	4	884920949
138	26	5	879024232
167	232	1	892738341
60	427	5	883326620
57	304	5	883698581
223	274	4	891550094
189	512	4	893277702
243	15	3	879987440
92	1049	1	890251826
246	416	3	884923047
194	165	4	879546723
241	690	2	887249482
178	248	4	882823954
254	1444	3	886475558
293	5	3	888906576
127	229	5	884364867
225	237	5	879539643
299	229	3	878192429
225	480	5	879540748
276	54	3	874791025
291	144	5	874835091
222	366	4	878183381
267	518	5	878971773
42	403	3	881108684
11	111	4	891903862
95	625	4	888954412
8	338	4	879361873
162	25	4	877635573
87	1016	4	879876194
279	154	5	875296291
145	275	2	885557505
119	1153	5	874781198
62	498	4	879373848
62	382	3	879375537
28	209	4	881961214
135	23	4	879857765
32	294	3	883709863
90	382	5	891383835
286	208	4	877531942
293	685	3	888905170
216	144	4	880234639
166	328	5	886397722
250	496	4	878090499
271	132	5	885848672
160	174	5	876860807
265	118	4	875320714
198	498	3	884207492
42	96	5	881107178
168	151	5	884288058
110	307	4	886987260
58	144	4	884304936
90	648	4	891384754
271	346	4	885844430
62	21	3	879373460
279	832	3	881375854
237	514	4	879376641
94	789	4	891720887
128	485	3	879966895
298	317	4	884182806
44	195	5	878347874
264	200	5	886122352
194	385	2	879524643
72	195	5	880037702
222	750	5	883815120
250	264	3	878089182
41	265	3	890687042
224	245	3	888082216
82	135	3	878769629
262	1147	4	879791710
293	471	3	888904884
216	658	3	880245029
250	140	3	878092059
59	23	5	888205300
286	379	5	877533771
244	815	4	880605185
7	479	4	891352010
174	368	1	886434402
87	274	4	879876734
194	1211	2	879551380
82	1134	2	884714402
13	836	2	882139746
13	272	4	884538403
244	756	2	880605157
305	427	5	886323090
95	787	2	888954930
43	14	2	883955745
299	955	4	889502823
57	419	3	883698454
84	405	3	883452363
269	504	4	891449922
299	111	3	877878184
194	466	4	879525876
160	135	4	876860807
99	268	3	885678247
10	486	4	877886846
259	117	4	874724988
85	427	3	879456350
303	919	4	879467295
213	273	5	878870987
121	514	3	891387947
90	98	5	891383204
49	559	2	888067405
42	794	3	881108425
155	323	2	879371261
68	117	4	876973939
172	177	4	875537965
19	4	4	885412840
268	231	4	875744136
5	2	3	875636053
305	117	2	886324028
44	294	4	883612356
43	137	4	875975656
279	1336	1	875298353
80	466	5	887401701
254	164	4	886472768
298	281	3	884183336
279	1240	1	892174404
66	298	4	883601324
18	443	3	880130193
268	1035	2	875542174
99	79	4	885680138
13	98	4	881515011
26	258	3	891347949
7	455	4	891353086
222	755	4	878183481
200	673	5	884128554
119	328	4	876923913
213	172	5	878955442
276	322	3	874786392
94	1217	3	891723086
130	379	4	875801662
38	328	4	892428688
160	719	3	876857977
293	1267	3	888906966
26	930	2	891385985
130	216	4	875216545
92	1079	3	886443455
256	452	4	882164999
1	61	4	878542420
72	48	4	880036718
56	755	3	892910207

In [0]:
dbutils.fs.cp("file:/databricks/driver/covertype.csv", "dbfs:/FileStore/tmp/covertype.csv")
dbutils.fs.cp("file:/databricks/driver/kafka.csv", "dbfs:/FileStore/tmp/kafka.csv")
dbutils.fs.cp("file:/databricks/driver/treecover.csv", "dbfs:/FileStore/tmp/treecover.csv")
dbutils.fs.cp("file:/databricks/driver/u.data.csv", "dbfs:/FileStore/tmp/u.data.csv")
dbutils.fs.cp("file:/databricks/driver/u.item.csv", "dbfs:/FileStore/tmp/u.item.csv")

Out[ 3 ]: True

## Part 1: Reading and Parsing Data

### Question 1:  Code Challenge - Load a CSV

- Load the CSV file at `dbfs:/FileStore/tmp/nl/treecover.csv` into a DataFrame.
- Use Apache Spark to read in the data, assigned to the variable `treeCoverDF`.
- Your method to get the CSV file into Databricks isn't graded. We are only concerned with how you use Spark to parse and load the actual data. 
- Please use the `inferSchema` option.

In [0]:
treeCoverDF = spark.read.load("dbfs:/FileStore/tmp/treecover.csv",
                     format="csv", sep=",", inferSchema="true", header="true")

### Question 2:  Code Challenge - Print the Schema

Use Apache Spark to display the Schema of the `treeCoverDF` Dataframe.

In [0]:
treeCoverDF.printSchema()

root
-- Id: integer (nullable = true)
-- Elevation: integer (nullable = true)
-- Aspect: integer (nullable = true)
-- Slope: integer (nullable = true)
-- Horizontal_Distance_To_Hydrology: integer (nullable = true)
-- Vertical_Distance_To_Hydrology: integer (nullable = true)
-- Horizontal_Distance_To_Roadways: integer (nullable = true)
-- Horizontal_Distance_To_Fire_Points: integer (nullable = true)
-- Cover_Type: integer (nullable = true)
-- Soil_Type: integer (nullable = true)
-- Wilderness_Area: integer (nullable = true)
-- Hillshade: string (nullable = true)

### Question 3:  Code Challenge - Rows & Columns

Use Apache Spark to display the number of rows and columns in the DataFrame.

In [0]:

print(treeCoverDF.count(),len(treeCoverDF.columns))

15120 12

#Part 2: Analysis

### Question 4:  Code Challenge - Summary Statistics for a Feature

Use Apache Spark to answer these questions about the `treeCoverDF` DataFrame:
- What is the range - minimum and maximum - of values for the feature `elevation`?
- What are the mean and standard deviation of the feature `elevation`?

In [0]:
from pyspark.sql.functions import *
treeCoverDF.agg(min(col("elevation")), max(col("elevation"))).show()

treeCoverDF.agg(mean(col("elevation")), stddev(col("elevation"))).show()

+--------------+--------------+
min(elevation)|max(elevation)|
+--------------+--------------+
 1863| 3849|
+--------------+--------------+

+------------------+----------------------+
 avg(elevation)|stddev_samp(elevation)|
+------------------+----------------------+
2749.3225529100528| 417.67818734804985|
+------------------+----------------------+

### Answer #4:

- Min `elevation`: 1863
- Max `elevation`: 3849
- Mean `elevation`: 2749.3225529100528
- Standard Deviation of `elevation`: 417.67818734804985

### Question 5:  Code Challenge - Record Count

Use Apache Spark to answer the following question:
- How many entries in the dataset have an `elevation` greater than or equal to 2749.32 meters **AND** a `Cover_Type` of 1 or 2?

In [0]:
treeCoverDF.filter('elevation >= 2749.32 and cover_type = 1 or cover_type = 2').show()

+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
 Id|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Horizontal_Distance_To_Fire_Points|Cover_Type|Soil_Type|Wilderness_Area|Hillshade|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
 3| 2804| 139| 9| 268| 65| 3180| 6121| 2| 12| 1| 3pm|
 4| 2785| 155| 18| 242| 118| 3090| 6211| 2| 30| 1| 9am|
 6| 2579| 132| 6| 300| -15| 67| 6031| 2| 29| 1| 3pm|
 12| 2886| 151| 11| 371| 26| 5253| 4051| 2| 30| 1| 3pm|
 13| 2742| 134| 22| 150| 69| 3215| 6091| 2| 30| 1| 9am|
 22| 2880| 209| 17| 216| 30| 4986| 4323| 2| 30| 1| 9am|
 28| 2962| 148| 16| 323| 23| 5916| 3395| 2| 29| 1| 9am|
 29| 2811| 135| 1| 212| 30| 3670| 5643| 2| 12| 1| Noon|
 36| 2900| 45| 19| 242| 20| 5199| 4115| 2| 29| 1| 3pm|
 42| 2570| 346| 2| 0| 0| 331| 5745| 2| 29| 1| 3pm|
 45| 2678| 128| 5| 95| 23| 1660| 6546| 2| 12| 1| 3pm|
 62| 2952| 107| 11| 42| 7| 5845| 3509| 2| 29| 1| Noon|
 63| 2705| 90| 8| 134| 22| 2023| 6615| 2| 12| 1| 3pm|
 68| 2919| 13| 13| 90| 6| 5321| 4060| 1| 29| 1| Noon|
 69| 2740| 54| 6| 218| 42| 2287| 6686| 2| 12| 1| 3pm|
 70| 2640| 80| 8| 180| -2| 1092| 5866| 2| 29| 1| 9am|
 71| 2843| 166| 12| 242| 53| 4434| 4956| 2| 30| 1| Noon|
 72| 3008| 45| 14| 277| 10| 6371| 3036| 2| 29| 1| 3pm|
 73| 2893| 114| 16| 108| 30| 5066| 4340| 2| 29| 1| 9am|
 74| 2850| 6| 9| 0| 0| 4858| 4548| 2| 16| 1| Noon|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
only showing top 20 rows

### Question 6: Code Challenge - Compute a Percentage

Use Apache Spark to answer the following question:
- What percentage of entries with `Cover_Type` 1 or 2 have an `elevation` at or above 2749.32 meters?

In [0]:
filtercount = treeCoverDF.filter('elevation >= 2749.32 and cover_type = 1 or cover_type = 2').count()
totalcount = treeCoverDF.count()
print((filtercount*100)/totalcount)

28.30026455026455

### Question 7: Code Challenge - Visualize Feature Distribution

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart that helps visualize the distribution of different Wilderness Areas in our dataset

In [0]:
display(treeCoverDF.select('Wilderness_Area').distinct())

Wilderness_Area 1 3 4 2

### Question 8: Code Challenge - Visualize Average Elevation by Cover Type 

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart showing the average elevation of each cover type with string labels for cover type

**NOTE: you will need to match the integer values in the column `treeCoverDF.Cover_Type` to the string values in `dbfs:/FileStore/tmp/nl/covertype.csv` to retrieve the Cover Type Labels. It is recommended to use an Apache Spark join.**

In [0]:
treeCoverType = spark.read.load("dbfs:/FileStore/tmp/covertype.csv",
                     format="csv", sep=",", inferSchema="true", header="true")

treeCoverType.show()   

newdf = treeCoverDF.join(treeCoverType,treeCoverDF.Cover_Type == treeCoverType.cover_type_key)

display(newdf.select('elevation','cover_type_label'))

elevation cover_type_label 2596 Aspen 2590 Aspen 2804 Lodgepole Pine 2785 Lodgepole Pine 2595 Aspen 2579 Lodgepole Pine 2606 Aspen 2605 Aspen 2617 Aspen 2612 Aspen 2612 Aspen 2886 Lodgepole Pine 2742 Lodgepole Pine 2609 Aspen 2503 Aspen 2495 Aspen 2610 Aspen 2517 Aspen 2504 Aspen 2503 Aspen 2501 Aspen 2880 Lodgepole Pine 2768 Aspen 2511 Aspen 2507 Aspen 2492 Aspen 2489 Aspen 2962 Lodgepole Pine 2811 Lodgepole Pine 2739 Aspen 2703 Aspen 2522 Aspen 2519 Aspen 2516 Aspen 2515 Aspen 2900 Lodgepole Pine 2709 Aspen 2511 Aspen 2749 Aspen 2686 Aspen 2699 Spruce/Fir 2570 Lodgepole Pine 2533 Aspen 2703 Aspen 2678 Lodgepole Pine 2529 Aspen 2524 Aspen 2536 Aspen 2498 Aspen 2489 Aspen 2713 Aspen 2739 Spruce/Fir 2696 Spruce/Fir 2510 Aspen 2502 Aspen 2722 Spruce/Fir 2500 Aspen 2486 Aspen 2489 Aspen 2489 Aspen 2490 Aspen 2952 Lodgepole Pine 2705 Lodgepole Pine 2507 Aspen 2500 Aspen 2493 Aspen 2509 Aspen 2919 Spruce/Fir 2740 Lodgepole Pine 2640 Lodgepole Pine 2843 Lodgepole Pine 3008 Lodgepole Pine 2893 Lodgepole Pine 2850 Lodgepole Pine 2628 Lodgepole Pine 2864 Lodgepole Pine 2827 Lodgepole Pine 2529 Lodgepole Pine 2808 Spruce/Fir 2840 Lodgepole Pine 2795 Spruce/Fir 2746 Lodgepole Pine 2847 Spruce/Fir 2840 Spruce/Fir 2537 Lodgepole Pine 2860 Spruce/Fir 2818 Lodgepole Pine 2801 Lodgepole Pine 2791 Lodgepole Pine 2745 Lodgepole Pine 2514 Lodgepole Pine 2788 Lodgepole Pine 2562 Lodgepole Pine 3073 Lodgepole Pine 2978 Lodgepole Pine 2860 Spruce/Fir 3067 Lodgepole Pine 2804 Lodgepole Pine 2562 Lodgepole Pine 2567 Lodgepole Pine 2998 Lodgepole Pine 2684 Lodgepole Pine 2676 Lodgepole Pine 2863 Spruce/Fir 2574 Lodgepole Pine 2511 Lodgepole Pine 3067 Lodgepole Pine 2909 Lodgepole Pine 2863 Spruce/Fir 2567 Lodgepole Pine 3182 Spruce/Fir 2818 Spruce/Fir 2752 Lodgepole Pine 2751 Lodgepole Pine 2569 Lodgepole Pine 3195 Spruce/Fir 2982 Lodgepole Pine 2850 Spruce/Fir 2586 Lodgepole Pine 2837 Spruce/Fir 2723 Lodgepole Pine 3070 Lodgepole Pine 3057 Lodgepole Pine 2818 Lodgepole Pine 2847 Lodgepole Pine 2606 Lodgepole Pine 2893 Lodgepole Pine 2716 Lodgepole Pine 2644 Lodgepole Pine 2613 Lodgepole Pine 2929 Lodgepole Pine 2767 Lodgepole Pine 2991 Spruce/Fir 2939 Lodgepole Pine 2537 Lodgepole Pine 3119 Lodgepole Pine 2942 Spruce/Fir 2886 Spruce/Fir 2775 Lodgepole Pine 2490 Lodgepole Pine 3106 Spruce/Fir 2831 Lodgepole Pine 2837 Lodgepole Pine 2890 Lodgepole Pine 2824 Lodgepole Pine 2728 Lodgepole Pine 2962 Spruce/Fir 3142 Lodgepole Pine 2847 Lodgepole Pine 2536 Lodgepole Pine 3168 Spruce/Fir 2926 Spruce/Fir 2893 Lodgepole Pine 2687 Lodgepole Pine 3109 Lodgepole Pine 3119 Lodgepole Pine 2916 Spruce/Fir 2795 Lodgepole Pine 2555 Lodgepole Pine 2499 Lodgepole Pine 3027 Spruce/Fir 3096 Spruce/Fir 2916 Lodgepole Pine 2591 Lodgepole Pine 3185 Spruce/Fir 3021 Spruce/Fir 3083 Lodgepole Pine 2860 Lodgepole Pine 3008 Lodgepole Pine 3034 Spruce/Fir 2873 Lodgepole Pine 3182 Spruce/Fir 2998 Lodgepole Pine 2658 Lodgepole Pine 3175 Spruce/Fir 2952 Spruce/Fir 2622 Lodgepole Pine 2955 Lodgepole Pine 2837 Lodgepole Pine 3182 Spruce/Fir 3077 Lodgepole Pine 2755 Lodgepole Pine 3047 Spruce/Fir 2880 Lodgepole Pine 3221 Spruce/Fir 3182 Spruce/Fir 3064 Spruce/Fir 2827 Lodgepole Pine 2518 Lodgepole Pine 3050 Spruce/Fir 3107 Spruce/Fir 2995 Lodgepole Pine 2721 Lodgepole Pine 3182 Spruce/Fir 2724 Lodgepole Pine 2515 Lodgepole Pine 3017 Spruce/Fir 3077 Lodgepole Pine 2929 Lodgepole Pine 2735 Lodgepole Pine 2893 Lodgepole Pine 3231 Spruce/Fir 2560 Lodgepole Pine 3162 Spruce/Fir 2982 Spruce/Fir 2731 Lodgepole Pine 3172 Spruce/Fir 2916 Lodgepole Pine 2890 Lodgepole Pine 2704 Lodgepole Pine 3198 Spruce/Fir 2985 Lodgepole Pine 3011 Lodgepole Pine 3024 Lodgepole Pine 2763 Lodgepole Pine 3244 Spruce/Fir 2955 Spruce/Fir 3201 Spruce/Fir 2972 Spruce/Fir 2985 Lodgepole Pine 2539 Lodgepole Pine 3001 Lodgepole Pine 2952 Spruce/Fir 2962 Lodgepole Pine 3004 Spruce/Fir 2727 Lodgepole Pine 3004 Lodgepole Pine 2886 Lodgepole Pine 3024 Spruce/Fir 3034 Lodgepole Pine 2903 Lodgepole Pine 2926 Spruce/Fir 2942 Lo

#Part 3: Data Ingestion, Cleansing, and Transformations

## Instructions 

This is a multi-step, data pipeline question in which you need to achieve a few objectives to build a successful job.

### Data Sets

#### `u.data.csv`

- The full u data set, 100000 ratings by 943 users on 1682 items. 
- Each user has rated at least 20 movies.  
- Users and items are numbered consecutively from 1. 
- The data is randomly ordered. 
- This is a tab separated file consisting of four columns: 
   - user id 
   - movie id 
   - rating 
   - date (unix seconds since 1/1/1970 UTC)

#### Desired schema

- `user_id INTEGER`
- `movie_id INTEGER`
- `rating INTEGER`
- `date DATE `

#### `u.item.csv`

- This is a `|` separated file consisting of six columns:
   - movie id
   - movie title
   - release date
   - video release date
   - IMDb URL
   - genre
- movie ids in this file match movie ids in `u.data`.

#### Desired schema

- `movie_id INTEGER`
- `movie_title STRING`

### Question 9:  Code Challenge - Load DataFrames

Use Apache Spark to perform the following:
1. define the correct schemas for each Data Set to be imported as described above  
   **note:** 
      - for `u.data.csv`, `date` *must* be stored using `DateType` with the format `yyyy-MM-dd`
      - you may need to ingest `timestamp` data using `IntegerType`
      - be sure to drop unneccesary columns for `u.item.csv`
1. import the two files as DataFrames names `uDataDF` and `uItemDF` using the schemas you defined and these paths:
   - `dbfs:/FileStore/tmp/u.data.csv`
   - `dbfs:/FileStore/tmp/u.item.csv`
1. order the `uDataDF` DataFrame by the `date` column

**NOTE:** Please display the DataFrames, `uDataDF` and `uItemDF` after loading.

#### `uDataDF`

In [0]:
from pyspark.sql.functions  import date_format

uDataDF = spark.read.load("dbfs:/FileStore/tmp/u.data.csv",
                     format="csv", sep="\t", inferSchema="true", header="false")

uDataDF.printSchema()
uDataDF = uDataDF.withColumnRenamed("_c0", "user_id").withColumnRenamed("_c1", "movie_id").withColumnRenamed("_c2", "rating").withColumnRenamed("_c3", "date").withColumn("tsDate",from_unixtime("date")).withColumn("newdate",date_format('tsDate', "yyyy-MM-dd")).drop('tsDate','date')

uDataDF.sort(desc("newdate")).show()





root
-- _c0: integer (nullable = true)
-- _c1: integer (nullable = true)
-- _c2: integer (nullable = true)
-- _c3: integer (nullable = true)

+-------+--------+------+----------+
user_id|movie_id|rating| newdate|
+-------+--------+------+----------+
 189| 30| 4|1998-04-22|
 189| 496| 5|1998-04-22|
 189| 120| 1|1998-04-22|
 189| 520| 5|1998-04-22|
 189| 462| 5|1998-04-22|
 189| 526| 4|1998-04-22|
 159| 259| 4|1998-04-22|
 189| 20| 5|1998-04-22|
 189| 197| 5|1998-04-22|
 189| 694| 4|1998-04-22|
 189| 45| 3|1998-04-22|
 189| 751| 4|1998-04-22|
 189| 276| 3|1998-04-22|
 189| 1115| 4|1998-04-22|
 189| 253| 4|1998-04-22|
 189| 1| 5|1998-04-22|
 189| 480| 5|1998-04-22|
 189| 1060| 5|1998-04-22|
 189| 705| 4|1998-04-22|
 189| 91| 3|1998-04-22|
+-------+--------+------+----------+
only showing top 20 rows

#### `uItemDF`

In [0]:
uItemDF = spark.read.load("dbfs:/FileStore/tmp/u.item.csv",
                     format="csv", sep="|", inferSchema="true", header="false")

uItemDF = uItemDF.select("_c0","_c1","_c2","_c3","_c4","_c5").withColumnRenamed("_c0", "movie_id").withColumnRenamed("_c1", "movie_title").withColumnRenamed("_c2", "release_date").withColumnRenamed("_c3", "video_release_date").withColumnRenamed("_c4", "imdb_url").withColumnRenamed("_c5", "genre")

uItemDF.show()

+--------+--------------------+------------+------------------+--------------------+-----+
movie_id| movie_title|release_date|video_release_date| imdb_url|genre|
+--------+--------------------+------------+------------------+--------------------+-----+
 1| Toy Story (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 2| GoldenEye (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 3| Four Rooms (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 4| Get Shorty (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 5| Copycat (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 6|Shanghai Triad (Y...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 7|Twelve Monkeys (1...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 8| Babe (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 9|Dead Man Walking ...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 10| Richard III (1995)| 22-Jan-1996| null|http://us.imdb.co...| 0|
 11|Seven (Se7en) (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 12|Usual Suspects, T...| 14-Aug-1995| null|http://us.imdb.co...| 0|
 13|Mighty Aphrodite ...| 30-Oct-1995| null|http://us.imdb.co...| 0|
 14| Postino, Il (1994)| 01-Jan-1994| null|http://us.imdb.co...| 0|
 15|Mr. Holland's Opu...| 29-Jan-1996| null|http://us.imdb.co...| 0|
 16|French Twist (Gaz...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 17|From Dusk Till Da...| 05-Feb-1996| null|http://us.imdb.co...| 0|
 18|White Balloon, Th...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 19|Antonia's Line (1...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 20|Angels and Insect...| 01-Jan-1995| null|http://us.imdb.co...| 0|
+--------+--------------------+------------+------------------+--------------------+-----+
only showing top 20 rows

### Question 10:  Code Challenge - Perform a Join

Use Apache Spark to do the following:
- join `uDataDF` and `uItemDf` on `movie_id` as a new DataFrame called `uMovieDF`  
   **note:** make sure you do not create duplicate `movie_id` columns
   
**NOTE:** Please display the DataFrame `uMovieDF`.

In [0]:
uMovieDF = uDataDF.join(uItemDF,['movie_id'])
uMovieDF.show()

+--------+-------+------+----------+--------------------+------------+------------------+--------------------+-----+
movie_id|user_id|rating| newdate| movie_title|release_date|video_release_date| imdb_url|genre|
+--------+-------+------+----------+--------------------+------------+------------------+--------------------+-----+
 242| 196| 3|1997-12-04| Kolya (1996)| 24-Jan-1997| null|http://us.imdb.co...| 0|
 302| 186| 3|1998-04-04|L.A. Confidential...| 01-Jan-1997| null|http://us.imdb.co...| 0|
 377| 22| 1|1997-11-07| Heavyweights (1994)| 01-Jan-1994| null|http://us.imdb.co...| 0|
 51| 244| 2|1997-11-27|Legends of the Fa...| 01-Jan-1994| null|http://us.imdb.co...| 0|
 346| 166| 1|1998-02-02| Jackie Brown (1997)| 01-Jan-1997| null|http://us.imdb.co...| 0|
 474| 298| 4|1998-01-07|Dr. Strangelove o...| 01-Jan-1963| null|http://us.imdb.co...| 0|
 265| 115| 2|1997-12-03|Hunt for Red Octo...| 01-Jan-1990| null|http://us.imdb.co...| 0|
 465| 253| 5|1998-04-03|Jungle Book, The ...| 01-Jan-1994| null|http://us.imdb.co...| 0|
 451| 305| 3|1998-02-01| Grease (1978)| 01-Jan-1978| null|http://us.imdb.co...| 0|
 86| 6| 3|1997-12-31|Remains of the Da...| 01-Jan-1993| null|http://us.imdb.co...| 0|
 257| 62| 2|1997-11-12| Men in Black (1997)| 04-Jul-1997| null|http://us.imdb.co...| 0|
 1014| 286| 5|1997-11-17|Romy and Michele'...| 25-Apr-1997| null|http://us.imdb.co...| 0|
 222| 200| 5|1997-10-05|Star Trek: First ...| 22-Nov-1996| null|http://us.imdb.co...| 0|
 40| 210| 3|1998-03-27|To Wong Foo, Than...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 29| 224| 3|1998-02-21|Batman Forever (1...| 01-Jan-1995| null|http://us.imdb.co...| 0|
 785| 303| 3|1997-11-14| Only You (1994)| 01-Jan-1994| null|http://us.imdb.co...| 0|
 387| 122| 5|1997-11-11|Age of Innocence,...| 01-Jan-1993| null|http://us.imdb.co...| 0|
 274| 194| 2|1997-11-14| Sabrina (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 1042| 291| 4|1997-09-21| Just Cause (1995)| 01-Jan-1995| null|http://us.imdb.co...| 0|
 1184| 234| 2|1998-04-08|Endless Summer 2,...| 01-Jan-1994| null|http://us.imdb.co...| 0|
+--------+-------+------+----------+--------------------+------------+------------------+--------------------+-----+
only showing top 20 rows

### Question 11:  Code Challenge - Perform an Aggregation

Use Apache Spark to do the following:
1. create an aggregate DataFrame, `aggDF` by
  1. extracting the year from the `date` (of the review)
  1. getting the average rating of each film per year as a column named `average_rating`
  1. ordering descending by year and average rating
1. write the resulting dataframe to a table named "movie_by_year_average_rating" in the Default database  
   **note:** use `mode(overwrite)` 

#### Desired Schema
The schema of you resulting DataFrame should be:
- `year INTEGER`
- `movie_title STRING`
- `average_rating DOUBLE`

**NOTE:** Please display the DataFrame `aggDF`.

In [0]:
from pyspark.sql.functions import year
from pyspark.sql.window import Window
uMovieDF1 = uMovieDF.withColumn("year",year("newdate"))


windowSpec = Window.partitionBy(uMovieDF1['year'])
cols = ["year","movie_title"]


uMovieDF1.withColumn('average_rating', avg("rating").over(Window.partitionBy(cols))).drop("movie_id","user_id","rating","newdate","release_date","video_release_date","imdb_url","genre").distinct().l̥show()
  

+--------------------+----+------------------+
 movie_title|year| average_rating|
+--------------------+----+------------------+
Kicking and Screa...|1997|3.7142857142857144|
 M*A*S*H (1970)|1997| 3.972972972972973|
 Paris, Texas (1984)|1997|3.7222222222222223|
White Man's Burde...|1997| 1.0|
 Passion Fish (1992)|1998|3.1176470588235294|
 Sleepover (1995)|1998| 3.0|
Spanish Prisoner,...|1998| 4.0|
 Thinner (1996)|1998|2.5555555555555554|
Trainspotting (1996)|1998| 3.802325581395349|
What's Love Got t...|1998|3.5185185185185186|
Anna Karenina (1997)|1997|3.0526315789473686|
 Big Bully (1996)|1997| 2.2|
 Kids (1995)|1997|2.8857142857142857|
Lost World: Juras...|1997| 2.9|
3 Ninjas: High No...|1998| 1.0|
Addiction, The (1...|1998| 3.0|
 Cronos (1992)|1998| 3.0|
Fille seule, La (...|1998| 3.5|
 GoldenEye (1995)|1998| 3.081967213114754|
I Love Trouble (1...|1998| 2.4|
+--------------------+----+------------------+
only showing top 20 rows

## Part 4: Fun with JSON

JSON values are typically passed by message brokers such as Kafka or Kinesis in a string encoding. When consumed by a Spark Structured Streaming application, this json must be converted into a nested object in order to be used.

Below is a list of json strings that represents how data might be passed from a message broker.

**Note:** Make sure to run the cell below to retrieve the sample data.

In [0]:
%python


sampleJson =  [{
 		"user": 100,
 		"ips": ["191.168.192.101", "191.168.192.103", "191.168.192.96", "191.168.192.99"]
 	},
 	{
 		"user": 101,
 		"ips": ["191.168.192.102", "191.168.192.105", "191.168.192.103", "191.168.192.107"]
 	},
 	{
 		"user": 102,
 		"ips": ["191.168.192.105", "191.168.192.101", "191.168.192.105", "191.168.192.107"]
 	},
 	{
 		"user": 103,
 		"ips": ["191.168.192.96", "191.168.192.100", "191.168.192.107", "191.168.192.101"]
 	},
 	{
 		"user": 104,
 		"ips": ["191.168.192.99", "191.168.192.99", "191.168.192.102", "191.168.192.99"]
 	},
 	{
 		"user": 105,
 		"ips": ["191.168.192.99", "191.168.192.99", "191.168.192.100", "191.168.192.96"]
 	}
 ]

### Question 12:  Code Challenge - Count the IPs

Use any coding techniques known to you to parse this list of JSON strings to answer the following question:
- how many occurrences of each IP address are in this list?

#### Desired Output
Your results should be this:


| ip | count |
|:-:|:-:|
| `191.168.192.96` | `3` |
| `191.168.192.99` | `6` |
| `191.168.192.100` | `2` |
| `191.168.192.101` | `3` |
| `191.168.192.102` | `2` |
| `191.168.192.103` | `2` |
| `191.168.192.105` | `3` |
| `191.168.192.107` | `3` |

**NOTE:** The order of your results is not important.

In [0]:
jsondf = spark.read.option('multiLine','true').json(sc.parallelize(sampleJson))
jsondf.select(explode(col("ips"))).groupBy("col").agg(count('col')).withColumnRenamed("col","ip").show()


+---------------+----------+
 ip|count(col)|
+---------------+----------+
 191.168.192.96| 3|
191.168.192.105| 3|
191.168.192.101| 3|
191.168.192.103| 2|
191.168.192.107| 3|
 191.168.192.99| 6|
191.168.192.100| 2|
191.168.192.102| 2|
+---------------+----------+

## Part 5: The Databricks API

### Question 13: Conceptual Question - the Databricks API

In 4-5 sentences, please explain what the Databricks API is used for at a high-level.

### Answer:

`The databricks API can be used for multiple purposes `

- Clusters API is used for managing the databricks spark clusters e.g. create, edit,. start, restart,delete and other cluster lifecyle calls`
- DBFS API is used to interact with various data sources`
- Groups API is used to manage groups of users.`
- Likewise there are more API like Jobs, ML FLow,SCIM, Secrets, Libraries,Token ,Workspace to interact with various entities of Databricks`

### Question 14: Conceptual Question - Explain an API Call

In 4-5 sentences, please explain what this API call. Be sure to discuss some key attributes about the cluster.

```
$ curl -n -X POST -H 'Content-Type: application/json'                      \
  -d '{                                                                     \
  "cluster_name": "high-concurrency-cluster",                               \
  "spark_version": "4.2.x-scala2.11",                                       \
  "node_type_id": "i3.xlarge",                                              \
  "spark_conf":{                                                            \
        "spark.databricks.cluster.profile":"serverless",                    \
        "spark.databricks.repl.allowedLanguages":"sql,python,r"             \
     },                                                                     \
     "aws_attributes":{                                                     \
        "zone_id":"us-west-2c",                                             \
        "first_on_demand":1,                                                \
        "availability":"SPOT_WITH_FALLBACK",                                \
        "spot_bid_price_percent":100                                        \
     },                                                                     \
   "custom_tags":{                                                          \
        "ResourceClass":"Serverless"                                        \
     },                                                                     \
       "autoscale":{                                                        \
        "min_workers":1,                                                    \
        "max_workers":2                                                     \
     },                                                                     \
  "autotermination_minutes":10                                              \
}' https://dogfood.staging.cloud.databricks.com/api/2.0/clusters/create '
```

### Answer:

`The above lines of code is used to create a high concurrency cluster, a high concurrency cluster is a managed cloud resource which provide Spark-native fine-grained sharing for maximum resource utilization and minimum query latencies. This type of cluster supports:
Preemption: Proactively preempts Spark tasks from over-committed users to ensure all users get their fair share of cluster time and their jobs complete in a timely manner even when contending with dozens of other users. This uses Spark Task Preemption for High Concurrency.
Fault isolation: Creates an environment for each notebook, effectively isolating them from one another.`

## Part 6: Security

### Question 15: Conceptual Question - Security on Databricks

Using the Databricks Documentation, what would you recommend to a Databricks and AWS customer for **securely** storing and accessing their data.

### Answer:

`Network Isolation - create a dedicated VPC/VNET in the customer’s account that only contains Databricks infrastructure. This ensures that all Databricks policies and controls have no access to other production infrastructure in a customers account. Additionally, all communication with Databricks’ control infrastructure goes through direct links (while additionally eliminating any Public IPs). This prevents any traffic from traversing the public internet and is also encrypted with mutual TLS v1.2.  And finally, multiple security groups are used so that if ever Databricks services need to connect to other customer VPC/VNETs (e.g., containing another production data source), there is no risk of exposing access to those services to the outside world.

Enable SSO - Integration with SSO identity providers (such as Okta, Onelogin etc.) and support for SCIM allow you to easily manage and secure users.

Enable ACLs - Cluster, Data, job, Workspace level

Enable encryption at rest(storage level) and in motion(using custom code).`

# This is the end of the official test. Bonus below!

## Part 7: Bonus: Data Science & Machine Learning

### Question 16: Conceptual Question - A Skewed Feature

One of these lines is the *mean* of this feature. The other is the *median*. Which of these lines is the **mean** - the red line or the black line?

<img width=400px src=https://www.evernote.com/l/AAEycL6CQ0hLi5V5pIo91Ko-Pfk2i0AnGyMB/image.png>

In [0]:
`The mean is always greater than the median, hence, red line is the median and black is the median.`

### Answer:

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`
%md

### Question 17: Conceptual Question - Exploratory Data Analysis

The plots below show the distribution of home selling prices differentiated by a few categorical features. Based on these plots, **which of these categorical features** - Property Type, Exterior Quality, or Month Sold - would you expect to be most associated with Price? **Why**?

<img width=1600px src=https://www.evernote.com/l/AAHulkcc20hHSJV6D1udKiwSDCN0S6oV_5YB/image.png>

### Answer:

`The Exterior quality feature seems to be most associated with price as we can see the given graph`

### Question 18: Conceptual Question - Analyze Model Performance

Consider the following results for a decision tree model against training and testing data sets:

`decision tree regression - train r2 score: 0.9944`  
`decision tree regression - test r2 score:  0.3119`


What is your assessment of this model?

### Answer:

`The model is highly over-fitted because the training accuracy is very high as compared to the testing accuracy, we need to retrain the model and do cross validation techniques to verify our accuracy`

### Question 19: Conceptual Question - Model Selection

A series of models has been built using the same training data, but each with a subset of features.

Consider the following results for a series of logistic regression models and then answer this question:

- Which model would you choose and why?
- What other things would you want to look at and why?

| model number | feature subset | logistic regression test accuracy|
|:-:|:-:|:-:|
| 1| feat_1 |	 0.631|
| 2| feat_2 |	 0.552|
| 3| feat_3 |	 0.868|
| 4| feat_4 |	 0.868|
| 5| feat_1, feat_2 |	 0.657|
| 6| feat_1, feat_3 |	 0.947|
| 7| feat_1, feat_4 |	 0.921|
| 8| feat_2, feat_3 |	 0.947|
| 9| feat_2, feat_4 |	 0.973|
| 10| feat_3, feat_4 |	 0.947|
| 11| feat_1, feat_2, feat_3 |	 0.947|
| 12| feat_1, feat_2, feat_4 |	 0.947|
| 13| feat_1, feat_3, feat_4 |	 0.947|
| 14| feat_2, feat_3, feat_4 |	 0.973|
| 15| feat_1, feat_2, feat_3, feat_4 |	 0.973|

### Answer:

`I would chose model number 14 because it is giving highest accuracy and also it is better than model number 15 because it has an extra feature less which is not affecting the accuracy at all`

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>